### PR TITLE
Clean up and better interfaces for multi-list support.

### PR DESCRIPTION
**Pull request by busykai**
1. Change C array to std::span.
2. Clean up docs and code.

### CODE REVIEW

File: framework/scripts/generate_test_list.py [hunk 1]
```diff
@@ -74,28 +74,32 @@ def main():
...
         for name, tests in test_lists.items():
-            f.write(f'static constexpr struct test *{name}_test_list[] = {{')
+            f.write(f'static constexpr struct test * const {name}_test_list_[] = {{')
             test_list = [ f'    &_test_{test}' for test in tests ]
             f.write(',\n'.join(test_list))
             f.write('\n};\n\n')
+            f.write(f'static constinit const span<struct test * const> {name}_test_list({name}_test_list_);\n\n');
         # write selector function
-        f.write('struct test * const *get_test_list(const char *platform, size_t *size) {\n')
+        f.write('const span<struct test * const> *get_test_list(const char *test_list_name) {\n')
```
Comment (thiagomacieira) [inline framework/scripts/generate_test_list.py:90]: Don't return a pointer to the span. Just return the span.

The list array itself, above, can remain the unchanged. I think you simply return the array in the generated test function code and `std::span` will automatically get its size.

**Review by thiagomacieira — CHANGES_REQUESTED** [review]

File: framework/scripts/generate_test_list.py (hunk 1 above)
Comment (busykai) [reply framework/scripts/generate_test_list.py:90]: I need to know if the function returns nothing, hence returning the nullptr. What would you suggest instead?

Comment (busykai) [reply framework/scripts/generate_test_list.py]: Did it with optional which is basically the same as a pointer, but in more characters. :)

File: framework/scripts/generate_test_list.py [hunk 2]
```diff
@@ -55,45 +59,50 @@ def main():
...
-        f.write("struct test * const *get_test_list(const char *platform, size_t *size) {\n")
+        f.write('optional<const span<struct test * const>> get_test_list(const char *test_list_name) {\n')
         if default_name is not None:
-            f.write(f'    if (!platform) {{ *size = sizeof({default_name}_test_list)/sizeof(struct test *); return (struct test * const *)&{default_name}_test_list; }}\n')
+            f.write(f'    if (!test_list_name) return {default_name}_test_list;\n')
         for name in test_lists.keys():
-            f.write(f'    if  (!strcmp(platform, "{name}")) {{ *size = sizeof({name}_test_list)/sizeof(struct test *); return (struct test * const *)&{name}_test_list; }}\n')
-        f.write("    return nullptr;\n")
-        f.write("}\n")
+            f.write(f'    if  (!strcmp(test_list_name, "{name}")) return {name}_test_list;\n')
```
Comment (thiagomacieira) [inline framework/scripts/generate_test_list.py:94]: Minor coding style suggestion: don't declare the spans as `static constinit` variables. Just return the actual array directly in the `return` statement.

File: framework/scripts/generate_test_list.py [hunk 3]
```diff
@@ -55,45 +59,50 @@ def main():
...
+        f.write('    return {};\n')
```
Comment (thiagomacieira) [inline framework/scripts/generate_test_list.py:95]:
```c++
    return std::nullopt;
```

**Review by thiagomacieira — APPROVED** [review]

**Review by thiagomacieira — APPROVED** [review]

**Pull request merged**